File: day-2-program-alarm/typescript/intCode.ts

export const intCode = (codes: number[]) => {
    let i = 0;
    for (; i < codes.length; i += 4) {
        let [first, second, third] = [codes[codes[i + 1]], codes[codes[i + 2]], codes[i + 3]]
        switch (codes[i]) {
            case 1:
                codes[third] = first + second
                break;
            case 2:
                codes[third] = first * second
                break;
            case 99:
                break;
        }
    }
    return codes
}

File: day-2-program-alarm/goLang/intCode.go
package main

type intCode []int

func (ic intCode) posOneAndTwo(i int) (int, int) {
	return ic[ic[i+1]], ic[ic[i+2]]
}

func (ic intCode) run() []int {
	for i := 0; i < len(ic); i += 4 {
		switch ic[i] {
		case 1:
			first, second := ic.posOneAndTwo(i)
			ic[ic[i+3]] = first + second
		case 2:
			first, second := ic.posOneAndTwo(i)
			ic[ic[i+3]] = first * second
		case 99:
			break
		default:
		}
	}
	return ic
}
